tests/Mutator/ReturnValue/ArrayOneItemTest.php

```diff
@@ -0,0 +1,102 @@
+<?php
+/**
+ * This code is licensed under the BSD 3-Clause License.
+ *
+ * Copyright (c) 2017, Maks Rafalko
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Tests\Mutator\ReturnValue;
+
+use Infection\Tests\Mutator\AbstractMutatorTestCase;
+
+/**
+ * @internal
+ */
+final class ArrayOneItemTest extends AbstractMutatorTestCase
+{
+    /**
+     * @dataProvider provideMutationCases
+     */
+    public function test_mutator($input, $expected = null): void
+    {
+        $this->doTest($input, $expected);
+    }
+
+    public function provideMutationCases(): \Generator
+    {
+        yield 'It mutates when return typehint is not nullable array' => [
+            $this->getFileContent('mutates-not-nullable-array.php'),
+            <<<'PHP'
+<?php
+
+namespace ArrayOneItem_NotNullableArray;
+
+class Test
+{
+    public function getCollection() : array
+    {
+        $collection = [1, 2, 3];
+        return count($collection) > 1 ? array_slice($collection, 0, 1, true) : $collection;
```

Comment:
The fact that the mutation produces a conditional statement opens up the possibility of false positives. At the code execution time, depending on the size of the collection, the mutation may or may not effectively happened but it will also be counted as such.

Some collections have one element by design. For instance, all sorts of JSON-based API responses that have one top-level element:
```js
{
   "response": { /* ... */}
}
```

The code that consumes this API and the corresponding tests don't expect any other elements in the collection, therefore the mutation (that effectively doesn't change anything) passes.

Reply:
Agree. And this is where the user can disable mutator for particular line, method or completely for the project.

Do you have any suggestions here?

This mutator is useful for the opposite scenario, where we are working with e.g. Doctrine repositories and expect > 1 elements in the returned array.

Reply:
> This mutator is useful for the opposite scenario, where we are working with e.g. Doctrine repositories and expect > 1 elements in the returned array.

Do you mean that they expect more than one item but only check the first one? Shouldn't the `ArrayItemRemoval` mutation flag such tests?

> Do you have any suggestions here?

Judging by the documentation, this is the only mutation that generates a conditional, one branch of which is the same as the original code. I suggest to avoid that and make sure the code is mutated unconditionally (however it's still not clear how it's different from `ArrayItemRemoval`).

Reply:
> Shouldn't the `ArrayItemRemoval` mutation flag such tests?

`ArrayItemRemoval` mutates the raw arrays, not variables. When you literally have `['a' => 1, 2, 3, 'd' => 4]` for example.

This mutator, `ArrayOneItem`, mutates variables. That's the difference.

Reply:
Indeed, even item removal cannot be done unconditionally. You cannot remove it from an empty array. Still, throwing an exception on an unsuccessful mutation should help avoid false positives.

Reply:
> even item removal cannot be done unconditionally. You cannot remove it from an empty array

we [don't mutate ](https://github.com/infection/infection/pull/649/files#diff-871cc5d111cdcdea40d147449608faeeR55-R107)empty array with `ArrayItermRemoval1 mutator :) So there are no conditions.

> Still, throwing an exception on an unsuccessful mutation should help avoid false positives.

Do you have false-positives with `ArrayOneItem` mutator because of conditions? What is the real source code?

I'm open to change it to throw an exception, just need more information about real cases and then we can create issue and discuss farther / implement.

Reply:
> we [don't mutate ](https://github.com/infection/infection/pull/649/files#diff-871cc5d111cdcdea40d147449608faeeR55-R107)empty array with `ArrayItermRemoval` mutator :) So there are no conditions.

That's the thing. You can see if the array literal is empty at mutation time and avoid false positives but you cannot do the same with a variable.

> Do you have false-positives with `ArrayOneItem` mutator because of conditions? What is the real source code?

Yes. See https://github.com/morozov/kih/pull/13.

Reply:
I have put this into an issue so we can track this a bit better: https://github.com/infection/infection/issues/803